### PR TITLE
Add a python script to generate graphs

### LINKED ISSUE #28: Please provide script for generating graphs
https://github.com/daanx/mimalloc-bench/issues/2#issuecomment-504799185

### PR DESCRIPTION
**Pull request by jvoisin**
It looks like this, but is an interactive svg, courtesy of [pygal]( https://www.pygal.org):

xxx

This should close #28

### CODE REVIEW

**Comment by daanx** [comment]
Ah, nice little script; you need to check though for errors -- I forgot the regex for that but I can post that later -- the error line is output just before actual the benchmark line which can be confusing.

However, I am _very_ hesitant to actually merge this PR and I personally think we should _not_ take mimalloc-bench in this direction. I think the current benchmark suite is very useful for allocator research and I tried hard to find a diverse set of benchmarks that each test particular aspects (like `xmalloc-test` or `Larson`), or approximate real-world behavior (like `mstress`), or are real programs (like `redis` or `cfrac`). And you can see for example that `snmalloc` does well on object migration between threads with `xmalloc-test`.  However, these are also still "small" benchmarks and do not reflect some other very important aspects of allocation: for example: (1) worst-case latencies are not considered, and (2) fragmentation of long running processes, (3) huge workloads, etc. For mimalloc most work over the past 2 years has been in optimizing these 3 issues and sometimes to the detriment of mimalloc-bench suite times. I know that for example `jemalloc` and `tcmalloc` also work hard on these  points -- for example, `jemalloc` is not the top scoring allocator in this suite but Firefox has been using it for many years and you can find that much work has gone into optimizing memory fragmentation even it made things slower on small benchmarks. 

What I would like to avoid is something like the computer language benchmark game where these considerations are lost and people just look at the micro benchmarks scores. I would already feel better if we were able to include benchmarks that would test specifically for the 95% percentile worst-case latency, memory fragmentation for large workloads over time, huge server workloads, and perhaps even a "secure" benchmark that does double-frees, buffer overruns etc to test for hardened allocators. If that would be part of the suite, it would at least show more clearly that, say, secure mimalloc may be a bit slower on `cfrac` (lots of small allocations) but also does not corrupt memory on a double-free or buffer overrun. Or jemalloc not fragmenting memory on large workloads, etc. 

In short, I would like to keep `mimalloc-bench` as a tool for research into allocators but would like to avoid using it as tool to see which one is "best" (as it is just not suitable for that at this point in time since there are still many untested aspects that matter a lot in practice).


ps. when I discuss the results of these benchmarks I always add a disclaimer and explanation to keep the above points in mind, and discuss each benchmark and what it is measuring.

**Comment by jvoisin** [comment]
I understand your concerns, but:
- `bench.sh` is already providing data, I'm just adding convenient visualisation
- you provided graphs yourself in the README.md file
- the axis is logarithmic, so it should only highlight catastrophic performances, and not mild disparities
- the amount and diversity of benchmarks should prevent lapidary affirmations like "X is a better allocator than Y"
- I won't send PR to make use of github's static pages to automatically generate and publish those graphs: it's up to users to run their benchmarks, generate their images, and draw their own conclusions.

Can you please open issues about other possible benchmarks that you would like to have included? I'll be happy to help :)

**Comment by daanx** [comment]
> * `bench.sh` is already providing data, I'm just adding convenient visualisation

That is exactly my point -- I want to keep it slightly inconvenient :-)

> In short, I would like to keep mimalloc-bench as a tool for research into allocators but would like to avoid using it as tool to see which one is "best" (as it is just not suitable for that at this point in time since there are still many untested aspects that matter a lot in practice).

I can see your viewpoint of course. Perhaps we can setup a meeting and discuss it further.

**Comment by mjp41** [comment]
> I can see your viewpoint of course. Perhaps we can setup a meeting and discuss it further.
(ps. I could not find an email, can you contact me over email? "daan" at "effp.org" Thanks!)

I would be super keen to be on the call if it can be suitable for UK time to?

> and perhaps even a "secure" benchmark that does double-frees, buffer overruns etc to test for hardened allocators.

I have been trying to think about this.  I have been adding some hardening features to `snmalloc` and it would be really interesting to have some objective comparison here.  I find it really hard to know how to compare things.  A mildly weaker hardening can be way better for performance, but it is unclear if it loses security.

I think the graph looks very clear.  When I do benchmarking, I typically run the suite about 20 times, and then produce box and whisker plots as many of the benchmarks can be noisy.  So spotting regressions can be hard.

I have found the benchmark suite good for spotting regressions.  I haven't used to to try to sell `snmalloc`.  But I do use it to justify changes in PRs.
